### PR TITLE
Timm Models integration to Optimum-intel[Draft]

### PR DESCRIPTION
**Pull request by sawradip**
[Will be added Later]

### CODE REVIEW

File: optimum/intel/openvino/modeling_timm.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+    outputs = OrderedDict([("logits", {0: "batch_size"})])
+
+
+class TimmPreTrainedModel(PreTrainedModel):
+    config_class = TimmConfig
+    base_model_prefix = "timm"
+    main_input_name = "pixel_values"
+
+
+class TimmModel(TimmPreTrainedModel):
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_timm.py:71]: I don't think we need such a model hierarchy, with TimmPreTrainedModel, TimmModel, etc. At least Optimum-Intel is not the place for such classes. What is required is just to put the logic of model creation and prefetch into the .from_timm() method of OVModelForTimm. Moreover, OVModelForTimm itself is not needed since there is no intent to expose it. It can be just a utility function or class that is not inherited from HF hierarchy at all.

File: optimum/intel/openvino/modeling_timm.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+            `config.num_labels > 1` a classification loss is computed (Cross-Entropy).
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        logits = self.timm(
+            pixel_values,
+            return_dict=return_dict,
+        )
+
+        loss = None
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_timm.py:151]: Loss is not needed at all in the context of further export to OpenVINO.

File: optimum/intel/openvino/modeling_timm.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        if not return_dict:
+            return logits
+
+        return ImageClassifierOutput(
+            loss=loss,
+            logits=logits.last_hidden_state,
+        )
+
+
+class TimmImageProcessor(BaseImageProcessor, ImageFeatureExtractionMixin):
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_timm.py:185]: I didn't expect but this is great that you created it. I just wonder how much from the overall Timm preprocessing options this class supports and how we will validate it?

File: optimum/intel/openvino/modeling_timm.py (hunk 2 above)
Comment (sawradip) [reply optimum/intel/openvino/modeling_timm.py]: I have removed that section.

File: optimum/intel/openvino/modeling_timm.py (hunk 1 above)
Comment (sawradip) [reply optimum/intel/openvino/modeling_timm.py]: I have removed the hierarchy as much as possible, leaving a small placeholder model to contain the `timm` models.

**Comment by sawradip** [comment]
@AlexKoff88  I have removed a of unnecessary code, and abstractions. Please take a look and let me know your feedback.

File: optimum/intel/openvino/modeling_timm.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        kwargs["revision"] = revision
+
+        config_dict = load_model_config_from_hf(pretrained_model_name_or_path)[0]
+        config_dict["num_labels"] = config_dict.pop("num_classes")
+        config_dict["image_size"] = config_dict.get("input_size")[-1]
+
+        return cls.from_dict(config_dict, **kwargs)
+
+
+class TimmOnnxConfig(ViTOnnxConfig):
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_timm.py:59]: You used `ViTOnnxConfig` as a base class here. Does it mean that only ViT models from Timm will work here?

Comment (AlexKoff88) [reply optimum/intel/openvino/modeling_timm.py:59]: I would inherit from `VisionOnnxConfig` to make it safer.

File: tests/openvino/test_modeling.py [hunk 5]
```diff
@@ -619,6 +623,27 @@ def test_pipeline(self, model_arch):
         self.assertTrue(isinstance(outputs[0]["label"], str))
         gc.collect()
 
+    @parameterized.expand(TIMM_MODELS)
```
Comment (AlexKoff88) [inline tests/openvino/test_modeling.py:626]: I think we should have a test which:
- load a timm model from hub
- saves it in IR locally (.save_pretrained)
- loads IR and does inference (random data is ok)

Comment (sawradip) [reply tests/openvino/test_modeling.py:626]: A test `test_timm_save_and_infer` is added.

File: optimum/intel/openvino/modeling_timm.py (hunk 3 above)
Comment (sawradip) [reply optimum/intel/openvino/modeling_timm.py:127]: I also wonder. Though I have added and used the class, it would be better if we had any methodical way for validating.

File: optimum/intel/openvino/modeling_diffusion.py [hunk 6]
```diff
@@ -217,9 +217,6 @@ def _from_pretrained(
                     cls.config_name,
                 }
             )
-            ignore_patterns = ["*.msgpack", "*.safetensors", "*pytorch_model.bin"]
-            if not from_onnx:
-                ignore_patterns.extend(["*.onnx", "*.onnx_data"])
```
Comment (fxmarty) [inline optimum/intel/openvino/modeling_diffusion.py:222]: I think this bit was a fix from a previous PR to avoid downloading ONNX weights when the OV ones are available.

File: optimum/intel/openvino/modeling_timm.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+class TimmOnnxConfig(VisionOnnxConfig):
+    DEFAULT_TIMM_ONNX_OPSET = 13
+    outputs = OrderedDict([("logits", {0: "batch_size"})])
+    NORMALIZED_CONFIG_CLASS = NormalizedVisionConfig
+    MIN_TORCH_VERSION = version.parse("1.11")
+
+    @property
+    def inputs(self) -> Dict[str, Dict[int, str]]:
+        return {"pixel_values": {0: "batch_size", 1: "num_channels", 2: "height", 3: "width"}}
```
Comment (fxmarty) [inline optimum/intel/openvino/modeling_timm.py:96]: Is this to support the ONNX export of timm models? Do you think it would be useful to move it (later) in optimum main repo?

File: optimum/intel/openvino/modeling_timm.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+        logits = self.model(
+            pixel_values,
+        )
+
+        return ImageClassifierOutput(
+            logits=logits,
+        )
+
+
+class TimmImageProcessor(BaseImageProcessor, ImageFeatureExtractionMixin):
```
Comment (fxmarty) [inline optimum/intel/openvino/modeling_timm.py:127]: Could you put a comment here pointing at what this class is adapted / copied from (if it is)?

**Review by fxmarty — APPROVED** [review]
I left a few comments for tiny things to be fixed, but otherwise looks good to me, thank you!

It could be worth mentioning in the documentation as well that `OVModelForImageClassification` can handle timm models.

File: optimum/intel/openvino/modeling_timm.py (hunk 8 above)
Comment (AlexKoff88) [reply optimum/intel/openvino/modeling_timm.py:127]: @sawradip, can you please answer.

Comment (sawradip) [reply optimum/intel/openvino/modeling_timm.py:127]: It is adapted from ViTImageProcessor - https://github.com/huggingface/transformers/blob/main/src/transformers/models/vit/image_processing_vit.py 

I have added a comment.

File: optimum/intel/openvino/modeling_diffusion.py (hunk 6 above)
Comment (sawradip) [reply optimum/intel/openvino/modeling_diffusion.py:223]: Merged the main branch. Hope this is solved.

Comment (fxmarty) [reply optimum/intel/openvino/modeling_diffusion.py:223]: Have you pushed?

I still see this PR branch reverting the fix: https://github.com/sawradip/optimum-intel/blob/integrate_timm/optimum/intel/openvino/modeling_diffusion.py#L229

Comment (sawradip) [reply optimum/intel/openvino/modeling_diffusion.py]: Now its fixed.   Right?

Comment (fxmarty) [reply optimum/intel/openvino/modeling_diffusion.py]: It is, thank you!

**Comment by fxmarty** [comment]
> It could be worth mentioning in the documentation as well that OVModelForImageClassification can handle timm models.

@sawradip Would you like to precise the doc / add an example, or should I merge as is?

**Comment by sawradip** [comment]
I will add some refernce to the doc as well add some description to the PR.

**Comment by sawradip** [comment]
@fxmarty , I have done some additions in docs, as well as added description to this PR. If everything seems alright, you can merge.

**Comment by sawradip** [comment]
@fxmarty  There was one error with style, fixed it.

Any feedback?

**Review by fxmarty — APPROVED** [review]
LGTM!

File: setup.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
     "datasets>=1.4.0",
     "sentencepiece",
     "scipy",
+    "timm",
```
Comment (echarlaix) [inline setup.py:20]: I don't think `timm` should become a hard dependency
